### ISSUE TITLE
[Bug] Memory Usage when Backpropping wrt Model Inputs

### ISSUE THREAD

**Issue by wjmaddox** [cornellius-gp/gpytorch#1560]
# 🐛 Bug

Memory usage tends to increase when generating lots of new models and backpropagating wrt their inputs. 

Comes up in certain types of acquisition function optimization like Knowledge Gradient. But, the current qKnowledgeGradient method in botorch doesn't seem to have this issue due to usage of fantasy model. 

## To reproduce

** Code snippet to reproduce **
```python
import gpytorch
import torch

device = torch.device("cuda:5")
train_x = torch.randn(50, 3, device = device)
train_y = torch.randn(50, 1, device = device)

# We will use the simplest form of GP model, exact inference
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood = gpytorch.likelihoods.GaussianLikelihood()):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel())
    
    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

model = ExactGPModel(train_x, train_y.squeeze(-1)).to(device)

def make_faked_model_in_bad_way(inputs):
    partial_inputs = inputs[..., :3, :].clone().contiguous()
    model.eval()
    fantasized_targets = model(partial_inputs).rsample(torch.Size((256, )))
    catted_inputs = torch.cat((
        partial_inputs, 
        model.train_inputs[0].expand(partial_inputs.shape[:-2] + model.train_inputs[0].shape[-2:]
    )), dim=-2)
    catted_targets = torch.cat((
        fantasized_targets, 
        model.train_targets.expand(fantasized_targets.shape[:-1] + model.train_targets.shape[-1:])#.unsqueeze(-1)
    ), dim=-1)
    new_model = ExactGPModel(
        catted_inputs.expand(catted_targets.shape[:-1] + catted_inputs.shape[-2:]), 
        catted_targets.squeeze(-1)
    ).to(device)
    return new_model

# this just needs to be some fn of the posterior
def construct_loss_function(inputs, model):
    other_inputs = inputs[..., 3:, :].clone().contiguous()
    model.eval()
    posterior = model(other_inputs)
    return (posterior.mean / (1 + posterior.variance)).mean((-1, 0))

# basically what the inputs to Knowledge Gradient (one shot version) look like
candidates = torch.randn(5, 67, 3, device = device, requires_grad = True)

# define our acquisition fn
acquisition_fn = lambda x: construct_loss_function(x, make_faked_model_in_bad_way(x))

print("memory allocated: ", torch.cuda.memory_allocated(device) / (1024 ** 3)) 
# approximately zero, these are small tensors

optimizer = torch.optim.Adam([candidates], lr = 0.001)

for i in range(1000):
    if i % 50 == 0:
        print("memory allocated: ", torch.cuda.memory_allocated(device) / (1024 ** 3))
    
    loss = -acquisition_fn(candidates).sum()

    def closure():
        optimizer.zero_grad()
        loss.backward()
        return loss

    optimizer.step(closure)

# in this one loop, builds up slowly (and somewhat non-monotonically) to ~0.5 GB.

new_points = candidates[:, :3, :].detach().clone()
del candidates

print("memory allocated: ", torch.cuda.memory_allocated(device) / (1024 ** 3))
torch.cuda.empty_cache()
print("memory allocated: ", torch.cuda.memory_allocated(device) / (1024 ** 3))
# stays at about ~0.5 GB.
```

** Stack trace/error message **
See slow memory leak above.

## Expected Behavior

Would expect constant time here, especially after deleting the candidates.

Including `with gpytorch.settings.detach_test_caches(False):` actually breaks the entire loop with a double backward, suggesting something isn't getting cleaned.

## System information

**Please complete the following information:**
-gpytorch 1.4.0 (master)
-pytorch 1.8.0
- ubuntu 18.04

## Additional context

I'm reasonably confident that this is a function of the `train_inputs` and `train_targets` caches not getting cleaned properly, but am quite stuck in debugging. So a pointer would be fantastic.

**Comment by gpleiss** [comment]
:/ I'll try to look at this soon

**Comment by wjmaddox** [comment]
I'm still looking at it and will let you know what I find. Currently going through the `get_fantasy_model` code and trying to figure out what's causing the success of that method.

**Comment by wjmaddox** [comment]
After a lot of digging, it looks proximally like [this line](https://github.com/cornellius-gp/gpytorch/blob/4229e53ab9ce9a38036343758ddc580ebb5634c2/gpytorch/lazy/lazy_tensor.py#L792) somehow cleans out the CUDA cache properly. That is, if it's commented out or replaced with a concatenation done lazily, the memory seems to increase, rather than staying low and stable. No idea why this would be the case so far, except it seems like it has to be a contiguous? issue or something.
